### ISSUE TITLE
Limit matterbridge ability to ddos server ;-)

### ISSUE THREAD

**Issue by jospoortvliet** [nextcloud/spreed#4098]
As each user can create an unlimited number of rooms and set up an unlimited number of matter bridges, there's a risk of the server getting overwhelmed.

We see three possible solutions:
* Use cgroups to limit memory and CPU usage of each matter bridge
* Use nice to lower the prio of the matterbridges
* Limit the number of matter bridges per user

### cgroups
The first is very hard to configure as it depends a lot on the server in question (how many users, how much CPU is available etc) - a default setup might be unworkable on a Pi and still totally let users blow up the server on a large instance. Plus, it is very hard to limit this per user as the user starts bridges one by one - so if you want to have, say, only 1% cpu usage max per user, what cgroup settings do you give the 1st bridge? And the second?

### use nice
Nice does not limit the amount of cpu usage, but would make it harder for each bridge to hurt the system's responsiveness. It can still run out of memory, however, and depending on the scheduler, the impact is still possibly non-zero.
-> Done in #4120

### limit per user
The main downside of this is that matter bridges could misbehave and perhaps, based on settings, some matterbridges use far more memory than others (eg set up a room with ALL possible connections, that each have high traffic -> maybe 5 of those already take a server down..)

## so what to do?
I think the easiest is to go the 'limit per user' route, but combine it with nice. So nice all matter bridge processes to the minimum - this at least decreases the risk of issues. I would, in addition, also use cgroups by default **if** we could come up with a configuration setting that works across the full range of systems Nextcloud supports - from a pi to a huge cluster. Maybe just a memory limitation of, say, 50 mb? That would make the use of matterbridge quite safe even with some malicious users on the system.

If we have a configurable limit of 10 matter bridges/user by default, with all matter bridges niced to 19, and a c-group limit of 50 mb, than a single user can at most use 500mb ram and, at most, hurt OTHER low-nice processes (as in, the other matter bridges). That would be bad-ish, but not allow a user to take down a system.

**Comment by nickvergessen** [comment]
I don't think we should do anything about this. If the admin wants to use matterbridge reducing/limiting it will lead to missed messages and that is bad.
So if that because an issue they need to limit the room creation (already possible) or disable matterbridge all together.

Using `nice` is still something I would find useful.

> limit per user

In group rooms we don't even know which moderator set it up. So a later moderator touching the settings would take over? And if they are at their limit already they would be unable to change?

**Comment by eneiluj** [comment]
So we could systematically run `nice -n19 /.../matterbridge` if it's ok for you.

I agree the limit per user does not seem possible because there can be multiple moderators for a room.

**Comment by jospoortvliet** [comment]
> > limit per user
> 
> In group rooms we don't even know which moderator set it up. So a later moderator touching the settings would take over? And if they are at their limit already they would be unable to change?

Hmmm, I understand. Maybe we should create a global limit so a malicious user can't cause the entire server to get in trouble? I'm just looking for a way to ensure that even bigger servers can still use this feature without having to worry about bad users taking their entire thing down...

**Comment by nickvergessen** [comment]
Bigger servers in company like environments should just use this:

> So if that because an issue they need to limit the room creation (already possible)

**Comment by jospoortvliet** [comment]
> Bigger servers in company like environments should just use this:
> 
> > So if that because an issue they need to limit the room creation (already possible)

Ok, that's nice with students maybe, but in a real company, they will never limit most employees from being able to create groups, right? I mean, you can't really use Talk that way very well. So that might help the TU Berlin, if they are willing to limit the students from working together (which I'm not even sure about), but not a company that uses Nextcloud for their employees... And certainly not a service provider.

Ah well, I guess we don't have to fix this before the first release. But I bet the question will come, if companies start using this. Maybe they won't, I dunno. I certainly want it on our company cloud, at least ;-)

**Comment by gary-kim** [comment]
Maybe in the long run, we could also look into having one matterbridge instance handle all rooms. That should help mitigate the problem more as well.
